src/idpi/operators/curl.py

```diff
@@ -0,0 +1,136 @@
+    "right": {"m": slice(1, -1), "c": slice(1, -1), "p": slice(2, None)},
+}
+
+
+def stpt(stencil: str, diff_type: str | None = None) -> dict[str, slice]:
+    """Get stencil points.
+
+    This dictionary can be used for selecting values from a data array.
+    Consider the following example, where ``A`` is a data array. ::
+
+        B = A[stpt('ccc')] + A[stpt('cpm')]
+
+    Here we add for every point a in ``A`` its upper (y+1) neighbor, one level
+    below (z-1) and store the result in ``B``.
+    In this method we allow 3d neighbor stencils. Points on the border won't
+    be included.
+
+    The stencil definition string must have three characters, corresponding to
+    the 'x', 'y' and 'z' dimensions respectively.
+    The characters can be
+    - 'c': center (+/-0)
+    - 'm': below (-1)
+    - 'p': above (+1)
+
+    Depending on the ``diff_type`` (set with ``set_diff_type``), the stencils
+    will be interpreted differently.
+    For 'left', positive stencils (+1, 'p') will be mapped to the original point.
+    For 'right', negative stencils (-1, 'm') will be mapped to the original point.
+
+    Args:
+        stencil (str): A string defining the stencil.
+        diff_type (str): A string indicating the difference type.
+
+    """
+    if diff_type is None:
+        diff_type = DEFAULT_DIFF_TYPE
+    s = STENCILS[diff_type]
+    return {d: s[k] for d, k in zip("xyz", stencil)}
+
+
+class FDField:
+    """Syntaxic sugar to apply the stencil to a field."""
+
+    def __init__(self, field: xr.DataArray, diff_type: str = DEFAULT_DIFF_TYPE):
+        """Init FDField."""
+        self.field = field
+        self.diff_type = diff_type
+
+    def __getitem__(self, stencil: str):
+        return self.field[stpt(stencil, diff_type=self.diff_type)]
+
+    @classmethod
+    def from_fields(cls, *fields, diff_type: str = DEFAULT_DIFF_TYPE):
+        for field in fields:
+            yield cls(field, diff_type)
+
+
+def curl(
+    U: xr.DataArray,
+    V: xr.DataArray,
+    W: xr.DataArray,
+    HHL: xr.DataArray,
+    inv_dlon: float,
+    inv_dlat: float,
+    wk: float,
+    sqrtg_r_s: xr.DataArray,
+    lat: xr.DataArray,
+) -> tuple[xr.DataArray, xr.DataArray, xr.DataArray]:
+    """Compute the curl of the velocity field."""
+    r_earth_inv = 1 / const.earth_radius
+    deg2rad = np.pi / 180
+    acrlat = cast(xr.DataArray, 1 / (np.cos(lat * deg2rad) * const.earth_radius))
+    tgrlat = cast(xr.DataArray, np.tan(lat * deg2rad))
+
+    u, v, w, hhl = FDField.from_fields(U, V, W, HHL)
+
+    dzeta_dlam = (
+        0.25
+        * inv_dlon
+        * sqrtg_r_s
+        * ((hhl["pcc"] - hhl["mcc"]) + (hhl["pcp"] - hhl["mcp"]))
+    )
+
+    dzeta_dphi = (
+        0.25
+        * inv_dlat
+        * sqrtg_r_s
+        * ((hhl["cpc"] - hhl["cmc"]) + (hhl["cpp"] - hhl["cmp"]))
+    )
+
+    # compute weighted derivatives for FD
+
+    du_dphi = ((u["cpc"] + u["mpc"]) - (u["cmc"] + u["mmc"])) * 0.25 * inv_dlat
```

Comment:
I am not very convinced about the usage of this custom syntax. Also notice that sometimes we use it as u["cmc"]  while below we also use u("cmc"). 

I will suggest looking into more standard syntax or aligned with other projects.
* variant 1: 
```u[0,1,0]``` equals the syntax used here as ```u["cpc"]```
it is used also in numba:
https://numba.pydata.org/numba-doc/latest/user/stencil.html
and in gt4py (cartesian mode)
https://github.com/GridTools/gt4py/blob/main/examples/cartesian/demo_horizontal_diffusion.ipynb
* variant 2:
Use a shift operator (concept of gt4py)
```shift(J, 1)(u)``` is equivalent to ```u["cpc"]```
reference: https://hackmd.io/ueLehFeOR5yyOCtn2Qdr-A

* variant 3: 
Use an offset object (used in icon4py)
```u[JOff(1)]``` is equivalent ```u["cpc"]```
reference: 
https://github.com/C2SM/icon4py/blob/main/atm_dyn_iconam/src/icon4py/atm_dyn_iconam/mo_solve_nonhydro_stencil_04.py